[GLIB] Generate serialization for DMABuf classes

psaavedra opened the pull request:
#### 4b241532961e42451c9ef18ea8d68bc9a993a934
<pre>
[GLIB] Generate serialization for DMABuf classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=274328">https://bugs.webkit.org/show_bug.cgi?id=274328</a>

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).
* Added new serialization description files for `DMABufColorSpace`,
  `DMABufFormat`, `DMABufObject` and `DMABufReleaseFlag`.
* Removed custom encode/decode methods in `DMABufObject`.
* Fixed argument coder templates in `Encoder.h` and `Decoder.h` by
  removing unnecessary `void` parameters.
* Updated `CMakeLists.txt` in WebCore and WebKit to include DMABuf
  serialization files.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.serialization.in: Added.
* Source/WebCore/platform/graphics/gbm/DMABufFormat.h:
(WebCore::DMABufFormat::Plane::Plane):
* Source/WebCore/platform/graphics/gbm/DMABufFormat.serialization.in: Added.
* Source/WebCore/platform/graphics/gbm/DMABufObject.h:
(WebCore::DMABufObject::DMABufObject): Deleted.
(WebCore::DMABufObject::encode): Deleted.
(WebCore::DMABufObject::decode): Deleted.
* Source/WebCore/platform/graphics/gbm/DMABufObject.serialization.in: Added.
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h:
(WebCore::DMABufReleaseFlag::DMABufReleaseFlag):
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.serialization.in: Added.
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decode):
* Source/WebKit/Platform/IPC/Encoder.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b241532961e42451c9ef18ea8d68bc9a993a934

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52073 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31400 "Hash 4b241532 for PR 28725 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4487 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55346 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2785 "Hash 4b241532 for PR 28725 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54377 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37781 "Failed to checkout and rebase branch from PR 28725") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2493 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42378 "Passed tests") | [❌ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/59/builds/2785 "Hash 4b241532 for PR 28725 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54167 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/37781 "Failed to checkout and rebase branch from PR 28725") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/4487 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23449 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/37781 "Failed to checkout and rebase branch from PR 28725") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/4487 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/955 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/37781 "Failed to checkout and rebase branch from PR 28725") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/4487 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56943 "Built successfully") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27191 "Hash 4b241532 for PR 28725 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/2493 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49773 "Passed tests") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28428 "Hash 4b241532 for PR 28725 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/4487 "Hash 4b241532 for PR 28725 does not build (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48993 "Passed tests") | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29336 "Hash 4b241532 for PR 28725 does not build (failure)") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28169 "Hash 4b241532 for PR 28725 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h:
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2022 Metrological Group B.V.
- * Copyright (C) 2022 Igalia S.L.
+ * Copyright (C) 2024 Igalia S.L.
```
It's not technically needed to update copyright years but when we do, typically we use ranges:

```suggestion
 * Copyright (C) 2022-2024 Igalia S.L.
```

aperezdc commented on Source/WebCore/platform/graphics/gbm/DMABufFormat.h:
```diff
@@ -126,6 +126,12 @@ struct DMABufFormat {
             , verticalSubsampling(PlaneDefinitionType::verticalSubsampling)
         { }
 
+        Plane(const FourCC& fourcc, const unsigned& hsValue, const unsigned& vsValue)
```
The constructor used during deserialization could be made `private` by adding a `friend` statement like this:

```cpp
friend struct IPC::ArgumentCoder<Plane, void>;
```

as part of the `struct Plane` declaration.

aperezdc commented on Source/WebCore/platform/graphics/gbm/DMABufObject.h:
```diff
@@ -40,20 +40,6 @@
 namespace WebCore {
 
 struct DMABufObject {
-    WTF_MAKE_NONCOPYABLE(DMABufObject);
```
Why do we need to remove `WTF_MAKE_NONCOPYABLE()` here? As far as I tried, the autogenerated serialization code can deal with non-copyable objects.

csaavedra replied:
Also the generated code uses move semantics, so if this is getting copied somewhere, that should be fixed instead.

aperezdc replied:
Adding any non-default implicit move constructor disables the implicit move assignment operator, and I suspect that's why Pablo removed `WTF_MAKE_NONCOPYABLE()`; the correct solution is to bring back the default move assignment operator with:

```cpp
DMABufObject& operator=(DMABufObject&&) = default;
```

I have a version of this patch that addresses the review issues and keeps the `DMABufObject` class as non-copyable, but I will have to submit it as a different PR.

aperezdc commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -129,7 +129,7 @@ class Decoder {
     std::optional<T> decode()
     {
-        std::optional<T> t { ArgumentCoder<std::remove_cvref_t<T>, void>::decode(*this) };
+        std::optional<T> t { ArgumentCoder<std::remove_cvref_t<T>>::decode(*this) };
```
I'm surprised this is needed. While working on https://github.com/WebKit/WebKit/pull/27313 one of the WIP versions of the patch added a new `.serialization.in` and this did not need to be changed 🤔

psaavedra replied:
The signature of the encode methods autogenerated change a bit:

```
In file included from sources/wpewebkit/Source/WebKit/Shared/WebPreferencesStore.h:29,
                 from sources/wpewebkit/Source/WebKit/UIProcess/WebPreferences.h:31,
                 from wpewebkit-2.45.upstream-20240511/DerivedSources/WebKit/WebPreferencesFeatures.cpp:29:
sources/wpewebkit/Source/WebKit/Platform/IPC/Encoder.h: In member function 'IPC::Encoder& IPC::Encoder::operator<<(T&&)':
sources/wpewebkit/Source/WebKit/Platform/IPC/Encoder.h:80:46: error: wrong number of template arguments (2, should be 1)
   80 |         ArgumentCoder<std::remove_cvref_t<T, void>>::encode(*this, std::forward<T>(t));
      |                                              ^~~~
In file included from recipe-sysroot/usr/include/c++/13.2.0/bits/stl_pair.h:60,
                 from recipe-sysroot/usr/include/c++/13.2.0/bits/stl_algobase.h:64,
                 from recipe-sysroot/usr/include/c++/13.2.0/algorithm:60,
                 from sources/wpewebkit/Source/WebKit/WebKit2Prefix.h:60,
                 from <command-line>:
recipe-sysroot/usr/include/c++/13.2.0/type_traits:3442:11: note: provided for 'template<class _Tp> using std::remove_cvref_t = typename std::remove_cvref::type'
 3442 |     using remove_cvref_t = typename remove_cvref<_Tp>::type;
      |           ^~~~~~~~~~~~~~
sources/wpewebkit/Source/WebKit/Platform/IPC/Encoder.h:80:50: error: template argument 1 is invalid
   80 |         ArgumentCoder<std::remove_cvref_t<T, void>>::encode(*this, std::forward<T>(t));
      |                                                  ^~
sources/wpewebkit/Source/WebKit/Platform/IPC/Encoder.h:80:86: error: expression list treated as compound expression in initializer [-fpermissive]
   80 |         ArgumentCoder<std::remove_cvref_t<T, void>>::encode(*this, std::forward<T>(t));
      |                                                                                      ^
```

psaavedra replied:
However it seems this is an issue for the Apple ports:

```
In file included from /Volumes/Data/worker/macOS-Ventura-Release-Build-EWS/build/WebKitBuild/Release/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource40-mm.mm:3:
In file included from /Volumes/Data/worker/macOS-Ventura-Release-Build-EWS/build/Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:29:
/Volumes/Data/worker/macOS-Ventura-Release-Build-EWS/build/Source/WebKit/Platform/IPC/Encoder.h:80:9: error: too few template arguments for class template 'ArgumentCoder'
```

csaavedra replied:
I don't think you should do this change.

aperezdc reviewed (commented):
Patch looks nice overall, I only have a couple of doubts but those might be because I have not touched much the new autogenerated serialization stuff. That's why it would be better someone else checks this as well&mdash;maybe @csaavedra, as he has been doing serialization patches lately?

psaavedra committed:
[GLIB] Generate serialization for DMABuf classes
https://bugs.webkit.org/show_bug.cgi?id=274328

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).
* Added new serialization description files for `DMABufColorSpace`,
  `DMABufFormat`, `DMABufObject` and `DMABufReleaseFlag`.
* Removed custom encode/decode methods in `DMABufObject`.
* Fixed argument coder templates in `Encoder.h` and `Decoder.h` by
  removing unnecessary `void` parameters.
* Updated `CMakeLists.txt` in WebCore and WebKit to include DMABuf
  serialization files.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.serialization.in: Added.
* Source/WebCore/platform/graphics/gbm/DMABufFormat.h:
(WebCore::DMABufFormat::Plane::Plane):
* Source/WebCore/platform/graphics/gbm/DMABufFormat.serialization.in: Added.
* Source/WebCore/platform/graphics/gbm/DMABufObject.h:
(WebCore::DMABufObject::DMABufObject): Deleted.
(WebCore::DMABufObject::encode): Deleted.
(WebCore::DMABufObject::decode): Deleted.
* Source/WebCore/platform/graphics/gbm/DMABufObject.serialization.in: Added.
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h:
(WebCore::DMABufReleaseFlag::DMABufReleaseFlag):
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.serialization.in: Added.
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decode):
* Source/WebKit/Platform/IPC/Encoder.h:

csaavedra commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2837,6 +2837,11 @@ list(APPEND WebCore_PRIVATE_FRAMEWORK_HEADERS
 
     platform/graphics/InbandTextTrackPrivate.serialization.in
 
+    platform/graphics/gbm/DMABufColorSpace.serialization.in
+    platform/graphics/gbm/DMABufFormat.serialization.in
+    platform/graphics/gbm/DMABufObject.serialization.in
+    platform/graphics/gbm/DMABufReleaseFlag.serialization.in
```
These files preferrably should go to `WebKit/Shared/glib/` instead of `WebCore`. Also, you could just put all DMABuf definitions in a single file to avoid clutter (there is already one, merging that one too into a new `ArgumentCodersDMABuf.serialization.in`).

psaavedra replied:
Does this also apply to .h files?

csaavedra replied:
Not necessarily. I don't think you need to change those.

csaavedra commented on Source/WebCore/platform/graphics/gbm/DMABufColorSpace.serialization.in:
```diff
@@ -0,0 +1,31 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+header: <WebCore/DMABufColorSpace.h>
```
You shouldn't need the header if the class for which a coder is being generated has a header with the exact same name and `.h` extension. `header:` is used when there are additional headers that need to be pulled and/or if the header has a different name (in which case you can use `[CustomHeader]` to prevent the default to be added).

csaavedra commented on Source/WebCore/platform/graphics/gbm/DMABufFormat.h:
```diff
@@ -30,6 +30,10 @@
 #include <cstddef>
 #include <cstdint>
 
+namespace IPC {
+template<typename T, typename U> struct ArgumentCoder;
+}
+
```
You can simply include `<wtf/ArgumentCoder.h>` instead, but I don't think you need this, read below.

csaavedra commented on Source/WebCore/platform/graphics/gbm/DMABufFormat.h:
```diff
@@ -129,6 +133,15 @@ struct DMABufFormat {
         unsigned horizontalSubsampling { 0 };
         unsigned verticalSubsampling { 0 };
+
+    private:
+        Plane(const FourCC& fourcc, const unsigned& hsValue, const unsigned& vsValue)
+            : fourcc(fourcc)
+            , horizontalSubsampling(hsValue)
+            , verticalSubsampling(vsValue)
+        { }
+
+        friend struct IPC::ArgumentCoder<Plane, void>;
```
You shouldn't need to add an explicit constructor since this is a struct and the data members are public, so aggregate initialization should work. If you use the same order in the serialization declaration as they are declared in the struct, the generated code should be able to create a new instance with something like `Plane { WTFMove(arg1), WTFMove(arg2), WTFMove(arg3) }`.

psaavedra replied:
AFAIK, in c++, structs can indeed benefit from aggregate initialization if they meet certain conditions like **no user-provided or inherited constructors**, no private or protected non-static data members, no virtual functions, and no virtual, private, or protected base classes.

However, the struct Plane does not qualify for aggregate initialization because it includes a user-provided constructor.

``` cpp
    struct Plane {                                                                                                     
        Plane() = default;                                                                                             
    
        template<typename PlaneDefinitionType>                                                                         
        Plane(const PlaneDefinitionType&)                                                                              
            : fourcc(PlaneDefinitionType::fourcc)                                                                      
            , horizontalSubsampling(PlaneDefinitionType::horizontalSubsampling)                                        
            , verticalSubsampling(PlaneDefinitionType::verticalSubsampling)                                            
        { }   
...
```

csaavedra commented on Source/WebCore/platform/graphics/gbm/DMABufFormat.serialization.in:
```diff
@@ -0,0 +1,73 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+header: <WebCore/DMABufFormat.h>
+struct WebCore::DMABufFormat {
```
Same as above regarding `header:`.

csaavedra commented on Source/WebCore/platform/graphics/gbm/DMABufObject.serialization.in:
```diff
@@ -0,0 +1,38 @@
+# THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+header: <WebCore/DMABufObject.h>
+[CustomHeader] struct WebCore::DMABufObject {
```
Same as above regarding `header:`.

csaavedra commented on Source/WebKit/Platform/IPC/Encoder.h:
```diff
@@ -77,7 +77,7 @@ class Encoder final {
     Encoder& operator<<(T&& t)
     {
-        ArgumentCoder<std::remove_cvref_t<T>, void>::encode(*this, std::forward<T>(t));
+        ArgumentCoder<std::remove_cvref_t<T>>::encode(*this, std::forward<T>(t));
```
I don't think you should do this change.

csaavedra commented on Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.serialization.in:
```diff
@@ -0,0 +1,27 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+header: <WebCore/DMABufReleaseFlag.h>
+struct WebCore::DMABufReleaseFlag {
```
Same as above regarding `header:`.
